Dockerfile

```diff
@@ -1,111 +1,279 @@
-FROM localstack/java-maven-node-python
+ARG IMAGE_TYPE=full
 
-MAINTAINER Waldemar Hummer (waldemar.hummer@gmail.com)
-LABEL authors="LocalStack Contributors"
+# java-builder: Stage to build a custom JRE (with jlink)
+FROM python:3.8-slim-buster as java-builder
+ARG TARGETARCH
 
-ARG LOCALSTACK_BUILD_DATE
-ARG LOCALSTACK_BUILD_GIT_HASH
+# install OpenJDK 11
+RUN apt-get update && apt-get install -y openjdk-11-jdk-headless
 
-ENV LOCALSTACK_BUILD_DATE=${LOCALSTACK_BUILD_DATE}
-ENV LOCALSTACK_BUILD_GIT_HASH=${LOCALSTACK_BUILD_GIT_HASH}
+ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-${TARGETARCH}
+
+# reate a custom, minimized JRE via jlink
+RUN jlink --add-modules \
+# include required modules
+java.base,java.desktop,java.instrument,java.management,java.naming,java.scripting,java.sql,java.xml,jdk.compiler,\
+# jdk.unsupported contains sun.misc.Unsafe which is required by certain dependencies
+jdk.unsupported,\
+# add additional cipher suites
+jdk.crypto.cryptoki,\
+# add ability to open ZIP/JAR files
+jdk.zipfs,\
+# Elasticsearch 7+ crashes without Thai Segmentation support
+jdk.localedata --include-locales en,th \
+    --compress 2 --strip-debug --no-header-files --no-man-pages --output /usr/lib/jvm/java-11 && \
+  cp ${JAVA_HOME}/bin/javac /usr/lib/jvm/java-11/bin/javac && \
+  cp -r ${JAVA_HOME}/include /usr/lib/jvm/java-11/include && \
+  mv /usr/lib/jvm/java-11/lib/modules /usr/lib/jvm/java-11/lib/modules.bk; \
+  cp -r ${JAVA_HOME}/lib/* /usr/lib/jvm/java-11/lib/; \
+  mv /usr/lib/jvm/java-11/lib/modules.bk /usr/lib/jvm/java-11/lib/modules; \
+  rm -rf /usr/bin/java ${JAVA_HOME} && ln -s /usr/lib/jvm/java-11/bin/java /usr/bin/java
+
+
+
+# base: Stage which installs necessary runtime dependencies (OS packages, java, maven,...)
+FROM python:3.8-slim-buster as base
+ARG TARGETARCH
+
+# Install runtime OS package dependencies
+RUN apt-get update && \
+        # Install dependencies to add additional repos
+        apt-get install -y --no-install-recommends ca-certificates curl && \
+        # Setup Node 14 Repo
+        curl -sL https://deb.nodesource.com/setup_14.x | bash - && \
+        # Install Packages
+        apt-get update && \
+        apt-get install -y --no-install-recommends \
+            # Runtime packages (groff-base is necessary for AWS CLI help)
+            git make openssl tar pixz zip unzip groff-base iputils-ping nss-passwords \
+            # Postgres
+            postgresql postgresql-client postgresql-plpython3 \
+            # NodeJS
+            nodejs && \
+        apt-get clean && rm -rf /var/lib/apt/lists/*
+
+SHELL [ "/bin/bash", "-c" ]
+
+# install Docker (CLI only)
```

Comment:
We can look into removing the Docker CLI from here, actually (now that we have the Docker Python SDK fully integrated). Would probably save us another ~50MB in the final image 💪 - we can look into that post initial release.. /cc @thrau  @dfangl @alexrashed

Reply:
good point! some folks do rely on the legacy docker client AFAIK, but we can maybe make that part of the full image.